### PR TITLE
WIP: Add waiting for OpenAPI specs being imported before starting readyz

### LINKED ISSUE #104324: Resource not found error when api server is started but OpenAPI specs are not imported
#### What happened:
When api server starts, it does not wait all OpenAPI specs are imported as shown below;

- It starts OpenAPI aggregator in different goroutine https://github.com/kubernetes/kubernetes/blob/e7efe219f8a5fcdff9647d45639696048c1f6147/staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go#L337
- It starts readyz endpoint for getting requests https://github.com/kubernetes/kubernetes/blob/e7efe219f8a5fcdff9647d45639696048c1f6147/staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go#L326

However, since the OpenAPI specs are not guaranteed that they are all imported at least once, commands like `kubectl explain somecrd` throws `resource not found` error in here https://github.com/openshift/kubernetes-kubectl/blob/0c5d119ce86414ebbd725f56967987c9cf72e496/pkg/cmd/explain/explain.go#L154

#### What you expected to happen:
Api server's readiness probe endpoint should ensure that all OpenAPI specs are imported and commands like `kubectl explain` returns correct results.

#### How to reproduce it (as minimally and precisely as possible):
Restart API server and immediately run `kubectl explain examplecrd` command.

#### Environment:
- Kubernetes version (use `kubectl version`):
v1.21.2
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

### PR DESCRIPTION
**Pull request by ardaguclu**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
OpenAPI spec import process is executed in goroutine, meanwhile api server's readyz endpoint is started and api server starts to get requests. However, there is no guarantee that all the OpenAPI specs are imported when the api server is ready to serve. This causes that whenever api server starts and if OpenAPI spec import is not completed, commands like `kubectl explain somecrd` returns `resource not found error`. This PR waits all the OpenAPI specs are imported at least once before api server is ready to serve.

#### Which issue(s) this PR fixes:
Fixes #104324

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

**Comment by liggitt** [comment]
is this guaranteed to succeed if there are unavailable aggregated API servers?

**Comment by liggitt** [comment]
I think this can cause kube-apiserver to never become ready if there's an unavailable aggregated api server, which we don't want

/hold

**Comment by ardaguclu** [comment]
Thanks for review @liggitt and this is a good point. The idea is the same with https://github.com/kubernetes/kubernetes/blob/e95983be570fb0c95c8341ecadd33f5577afb9cf/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go#L234. Maybe we can allow failures while first import(best effort) and not block apiserver after that letting worker queue handles failures?.

**Comment by ardaguclu** [comment]
@liggitt after the latest changes, kube-aggregator services do not block apiserver anymore. Could you please take a look one last time?. Thank you.

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go [hunk 1]
```diff
@@ -76,13 +77,28 @@ func NewAggregationController(downloader *aggregator.Downloader, openAPIAggregat
...
+func (c *AggregationController) Run(stopCh <-chan struct{}, initialAggrDoneCh chan<- struct{}) {
 	defer utilruntime.HandleCrash()
 	defer c.queue.ShutDown()
 
 	klog.Info("Starting OpenAPI AggregationController")
 	defer klog.Info("Shutting down OpenAPI AggregationController")
 
+	// Initially aggregate OpenAPI spec before running worker.
+	var wg sync.WaitGroup
+	for _, name := range c.openAPIAggregationManager.GetAPIServiceNames() {
```
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go:89]: is this guaranteed to be after the informer has synchronized?

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go [hunk 2]
```diff
@@ -76,13 +77,28 @@ func NewAggregationController(downloader *aggregator.Downloader, openAPIAggregat
...
+		wg.Add(1)
+		go func(name string, wg *sync.WaitGroup) {
+			defer wg.Done()
+			if _, err := c.syncHandler(name); err != nil {
```
Comment (deads2k) [inline staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go:93]: I can't remember if this method is threadsafe per-key.

Comment (sttts) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go:93]: It's not a good idea, even if it is thread safe. I think it will trigger multiple spec aggregations this way.

Comment (sttts) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go:93]: hmm, actually not so sure. The existing code is hard to follow.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go [hunk 3]
```diff
@@ -334,7 +334,12 @@ func (s *APIAggregator) PrepareRun() (preparedAPIAggregator, error) {
 	// add post start hook before generic PrepareRun in order to be before /healthz installation
 	if s.openAPIConfig != nil {
 		s.GenericAPIServer.AddPostStartHookOrDie("apiservice-openapi-controller", func(context genericapiserver.PostStartHookContext) error {
-			go s.openAPIAggregationController.Run(context.StopCh)
+			initialOpenAPIAggrDoneCh := make(chan struct{})
+			go s.openAPIAggregationController.Run(context.StopCh, initialOpenAPIAggrDoneCh)
+			select {
+			case <-context.StopCh:
+			case <-initialOpenAPIAggrDoneCh:
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:341]: It is ensured that the APIService informer is ready here? Otherwise, we won't see all api service names.

I believe the `AggregationController.AddAPIService` func is called via informer events and it registers the api services asynchronously.

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go (hunk 1 above)
Comment (sttts) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go:89]: same question as mine above

File: staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go (hunk 2 above)
Comment (ardaguclu) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go:93]: GetAPIServiceInfo is thread safe https://github.com/kubernetes/kubernetes/blob/e95983be570fb0c95c8341ecadd33f5577afb9cf/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go#L337

UpdateApiServiceSpec is also thread safe https://github.com/kubernetes/kubernetes/blob/e95983be570fb0c95c8341ecadd33f5577afb9cf/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/aggregator.go#L279.

The only place I'm not sure is Downloader https://github.com/kubernetes/kubernetes/blob/e95983be570fb0c95c8341ecadd33f5577afb9cf/staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/downloader.go#L54

Comment (ardaguclu) [reply staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go:93]: > It's not a good idea, even if it is thread safe. I think it will trigger multiple spec aggregations this way.

@sttts maybe I can apply the method you use in here https://github.com/kubernetes/kubernetes/blob/e95983be570fb0c95c8341ecadd33f5577afb9cf/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery_controller.go#L217.

I did not use wait.Poll in here not to block apiserver.

File: staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go [hunk 4]
```diff
@@ -77,7 +77,7 @@ func NewController(crdInformer informers.CustomResourceDefinitionInformer) *Cont
 }
 
 // Run sets openAPIAggregationManager and starts workers
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go:79]: godoc openapiControllerSyncedCh

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go [hunk 5]
```diff
@@ -222,7 +222,12 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
 		// choose to not enable OpenAPI by having null openAPIConfig, and thus OpenAPIVersionedService
 		// and StaticOpenAPISpec are both null. In that case we don't run the CRD OpenAPI controller.
 		if s.GenericAPIServer.OpenAPIVersionedService != nil && s.GenericAPIServer.StaticOpenAPISpec != nil {
-			go openapiController.Run(s.GenericAPIServer.StaticOpenAPISpec, s.GenericAPIServer.OpenAPIVersionedService, context.StopCh)
+			openapiControllerSyncedCh := make(chan struct{})
+			go openapiController.Run(s.GenericAPIServer.StaticOpenAPISpec, s.GenericAPIServer.OpenAPIVersionedService, context.StopCh, openapiControllerSyncedCh)
+			select {
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:227]: a comment wouldn't harm `// wait for first sync to finish in order to serve a complete OpenAPI spec before ready`

File: staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go [hunk 6]
```diff
@@ -114,6 +114,7 @@ func (c *Controller) Run(staticSpec *spec.Swagger, openAPIService *handler.OpenA
 		utilruntime.HandleError(fmt.Errorf("failed to initially create OpenAPI spec for CRDs: %v", err))
 		return
 	}
+	close(openapiControllerSyncedCh)
```
Comment (sttts) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go:117]: does this mean aggregated APIServices are also synched? Or only CRDs?

Comment (ardaguclu) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go:117]: That is a good question. But I think, this is only for CRDs.

Comment (sttts) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/controller.go:118]: Can you check and add a comment?

File: staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go [hunk 7]
```diff
@@ -222,7 +222,13 @@ func (c completedConfig) New(delegationTarget genericapiserver.DelegationTarget)
 		// choose to not enable OpenAPI by having null openAPIConfig, and thus OpenAPIVersionedService
 		// and StaticOpenAPISpec are both null. In that case we don't run the CRD OpenAPI controller.
 		if s.GenericAPIServer.OpenAPIVersionedService != nil && s.GenericAPIServer.StaticOpenAPISpec != nil {
-			go openapiController.Run(s.GenericAPIServer.StaticOpenAPISpec, s.GenericAPIServer.OpenAPIVersionedService, context.StopCh)
+			// Wait for first sync to finish in order to serve a complete OpenAPI spec before ready
```
Comment (Jefftree) [inline staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:225]: Can I ask why only the CRD openapi controller is blocked on? Even if the openapi for CRDs is ready, it doesn't guarantee that the aggregator will have the aggregated openapi ready at the `openapi/v2` endpoint? (I might be misunderstanding something here since I remember your original approach directly modified the aggregator)

Comment (ardaguclu) [reply staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/apiserver.go:225]: @Jefftree, after discussing with @sttts, as you suspect there is still a race. Plausible option is to make this initial aggregation synchronous(from apiextension -> aggregator). Otherwise, there will always be race.

File: pkg/controlplane/controller/crdregistration/crdregistration_controller.go [hunk 8]
```diff
@@ -40,6 +40,8 @@ import (
 type AutoAPIServiceRegistration interface {
 	// AddAPIServiceToSync adds an API service to auto-register.
 	AddAPIServiceToSync(in *v1.APIService)
+	// AddAPIServiceToSyncSynchronously adds an API service to auto-register synchronously without using queue.
+	AddAPIServiceToSyncSynchronously(in *v1.APIService)
```
Comment (sttts) [inline pkg/controlplane/controller/crdregistration/crdregistration_controller.go:44]: that's an interesting idea

Comment (sttts) [reply pkg/controlplane/controller/crdregistration/crdregistration_controller.go:44]: Does it work out, i.e. get you get the wiring through without the race?

Comment (ardaguclu) [reply pkg/controlplane/controller/crdregistration/crdregistration_controller.go:44]: I'm still trying to find a way how to test the idea(currently could not find) and PR is still WIP but my theory is such that;
crdregistration_controller enqueue apiservices in here https://github.com/kubernetes/kubernetes/blob/c964fe6001376c8ad209656175bd3c4087ce5d71/pkg/controlplane/controller/crdregistration/crdregistration_controller.go#L210

However, autoregistration_controller does not start consuming queue until here unleashes https://github.com/kubernetes/kubernetes/blob/c964fe6001376c8ad209656175bd3c4087ce5d71/cmd/kube-apiserver/app/aggregator.go#L150

Meanwhile, apiserviceregistration controller starts consuming https://github.com/kubernetes/kubernetes/blob/c964fe6001376c8ad209656175bd3c4087ce5d71/staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go#L264
and aggregating OpenAPI specs for API services in here https://github.com/kubernetes/kubernetes/blob/c964fe6001376c8ad209656175bd3c4087ce5d71/staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go#L395

However, due to the race between autoregistration and apiserviceregistration controllers, some OpenAPI specs are being missed in first run.

File: staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go [hunk 9]
```diff
@@ -193,6 +197,7 @@ func (c completedConfig) NewWithDelegate(delegationTarget genericapiserver.Deleg
 		openAPIConfig:              c.GenericConfig.OpenAPIConfig,
 		egressSelector:             c.GenericConfig.EgressSelector,
 		proxyCurrentCertKeyContent: func() (bytes []byte, bytes2 []byte) { return nil, nil },
+		CrdApiRegistrationDoneCh:   make(chan struct{}),
```
Comment (ardaguclu) [inline staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:200]: I still have doubts about this is the correct way to order the controllers(autoregistration&apiserviceregistration).

File: pkg/controlplane/controller/crdregistration/crdregistration_controller.go (hunk 8 above)
Comment (ardaguclu) [reply pkg/controlplane/controller/crdregistration/crdregistration_controller.go:44]: @sttts, current change(crdregistration calls autoregister synchronously) does not work in OpenShift. As you worried about deadlock(Apiserver and SDN?), something blocks bootstrap completes installation.

**Comment by ardaguclu** [comment]
I think, this https://github.com/kubernetes/kubernetes/pull/104748 PR will also fix that problem.

In favor of this, I'm closing this PR.

/close
